compiler/rustc_parse/src/parser/item.rs

```diff
@@ -97,6 +88,65 @@ impl<'a> Parser<'a> {
         let mod_spans = ModSpans { inner_span: lo.to(self.prev_token.span), inject_use_span };
         Ok((attrs, items, mod_spans))
     }
+
+    /// Tries to parse the item as a statement to provide further diagnostics.
+    fn fallback_incorrect_item(&mut self) -> rustc_errors::Diag<'a> {
+        let token_str = super::token_descr(&self.token);
+        let token_span = self.token.span;
+        let mut err =
+            self.dcx().struct_span_err(token_span, format!("expected item, found {token_str}"));
+
+        let mut do_default_diag = true;
+
+        match self.parse_full_stmt(AttemptLocalParseRecovery::No) {
+            Ok(Some(stmt)) => {
+                do_default_diag = false;
+                let span = stmt.span;
+                match &stmt.kind {
+                    StmtKind::Let(_) => {
+                        err.span_label(span, "unexpected `let` binding outside of a function")
+                            .help(format!("consider using `const` or `static` instead of `let` for global variables, or put it inside of a function:  fn foo() {{ {} }}",
+                            pprust::stmt_to_string(&stmt)));
+                    }
+                    StmtKind::Semi(expr) => {
+                        err.span_label(span, "unexpected expression").help(format!(
+                            "consider putting it inside a function: fn foo() {{ {}; }}",
```

Comment:
This diagnostic can potentially become very big. Either use structured suggestions, or do not put a suggestion into the diagnostic message itself.